### PR TITLE
Enable `link-arg` link kind inside of `#[link]` attribute

### PR DESCRIPTION
**Pull request by zetanumbers**
https://github.com/rust-lang/rust/issues/99427#issuecomment-1234443468

> ...
> This would help to make `link-arg` usable in `#[link]` attributes and e.g. wrap libc and libgcc into a group (*) in the libc crate like
> 
> ```
> #[link(kind = "link-arg", name = "--start-group")]
> #[link(kind = "static", name = "c")]
> #[link(kind = "static", name = "gcc")]
> #[link(kind = "link-arg", name = "--end-group")]
> ```
> 
> (*) to address cyclic dependencies between them
> 
> This is an analogue of CMake's LINKER: prefix (https://cmake.org/cmake/help/git-stage/command/target_link_options.html#handling-compiler-driver-differences), and was discussed as a possible future extension in the link modifier RFC (https://github.com/rust-lang/rfcs/blob/master/text/2951-native-link-modifiers.md#support-linkarg--string-in-addition-to-the-modifiers).

### CODE REVIEW

File: tests/run-make/linker-flags-from-attribute-from-dep/Makefile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+include ../tools.mk
+
+all:
+	# Build deps
+	$(RUSTC) native_dep_1.rs --crate-type=staticlib
+	$(RUSTC) native_dep_2.rs --crate-type=staticlib
+	$(RUSTC) rust_dep.rs --crate-type=lib
+
+	# Check sequence of linker args
+	$(RUSTC) main.rs --extern lib=$(TMPDIR)/librust_dep.rlib --crate-type=bin --print link-args | $(CGREP) -e 'native_dep_1.*some_flag.*native_dep_2'
```
Comment (petrochenkov) [inline tests/run-make/linker-flags-from-attribute-from-dep/Makefile:10]: This is an adaptation of `tests\run-make\pass-linker-flags-from-dep\Makefile`, right?
Maybe it's better to just merge it into that test to keep the similar testing logic in the same makefile.

Same for `tests/run-make/linker-flags-from-attribute(-flavor)/Makefile`, which are adaptations of other tests.

**Comment by bjorn3** [comment]
I thought we had this ability in the past but removed it because it depends on the exact linker that is used.

**Comment by petrochenkov** [comment]
@bjorn3 
The previous version (`#[link_args]`) was removed in https://github.com/rust-lang/rust/pull/83820.
It was different than this one and more similar to the `-C link-arg` flag - its relative order with `-l` flags and `#[link]` attribute wasn't respected, an when used in a rlib it wasn't recorded in rlib metadata and wasn't later used during the final link.

This version is an attribute version of `-l link-arg`, so the order is respected, and it's preserved in metadata, and it otherwise behaves like a native library.
This new setup was described as a future possibility and potential replacement for `#[link_args]` in https://github.com/rust-lang/rfcs/blob/master/text/2951-native-link-modifiers.md#support-linkarg--string-in-addition-to-the-modifiers.

>depends on the exact linker that is used

The portability here is relatively good because `-Wl,` is added automatically if the current linker flavor requires it (`-verbatim` is the default).
There are also some relevant docs at `enum LinkerFlavor` - basically in practice for each given target there is a basic linker flavor with its own interface and two "modifiers" `cc` and `lld` that can be added to it. `lld` doesn't change the interface, and `cc` requires adding `-Wl,` which can be done automatically, so use cases like `--(start,end)-group` in https://github.com/rust-lang/rust/issues/99427#issuecomment-1234443468 should work.

**Review by petrochenkov — APPROVED** [review]

**Comment by petrochenkov** [comment]
r=me after squashing commits into one.
@rustbot author

**Comment by ZetaNumbers** [comment]
r=@petrochenkov

**Comment by ZetaNumbers** [comment]
r=petrochenkov

**Pull request merged**